pkg/subscriber/processor/goal_events_dwh.go

```diff
@@ -271,34 +304,66 @@ func (w *goalEvtWriter) linkGoalEventByExperiment(
 	if len(exps) == 0 {
 		return nil, false, ErrExperimentNotFound
 	}
-	evals := make([]*featureproto.Evaluation, 0, len(exps))
+	evals := make([]*ecstorage.UserEvaluation, 0, len(exps))
 	for _, exp := range exps {
 		// Get the user evaluation using the experiment info
-		ev, err := w.getUserEvaluation(
+		userID := getUserID(event.UserId, event.User)
+		eval, err := w.getUserEvaluation(
 			ctx,
-			event.User,
-			environmentId,
-			tag,
+			environmentID,
+			userID,
 			exp.FeatureId,
 			exp.FeatureVersion,
+			exp.StartAt,
+			exp.StopAt,
 		)
 		if err != nil {
-			if errors.Is(err, ErrEvaluationsAreEmpty) {
-				w.logger.Error("evaluations are empty",
+			if errors.Is(err, ecstorage.ErrNoResultsFound) {
+				w.logger.Error("Evaluation not found",
 					zap.Error(err),
-					zap.String("environmentId", environmentId),
+					zap.String("environmentId", environmentID),
 					zap.Any("goalEvent", event),
 				)
+				subscriberHandledCounter.WithLabelValues(subscriberGoalEventDWH, codeUserEvaluationNotFound).Inc()
+				if err := w.storeRetryMessage(&retryMessage{
+					GoalEvent:     event,
+					EnvironmentID: environmentID,
+					RetryCount:    0,
+					ID:            fmt.Sprintf("%s-%s-%d", event.GoalId, userID, event.Timestamp),
+				}); err != nil {
+					subscriberHandledCounter.WithLabelValues(subscriberGoalEventDWH, codeFailedToStoreRetryMessage).Inc()
+					w.logger.Error("Failed to store retry message",
+						zap.Error(err),
+						zap.String("environmentId", environmentID),
+						zap.Any("goalEvent", event),
+					)
+					return nil, true, err
```

Comment:
If it fails to store the message, it will not acknowledge the message from PubSub.